x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -160,13 +170,19 @@ public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDecider
                 unassignedBytes,
                 unassignedBytesUnassignedShards.shardIds(),
                 assignedBytes,
-                assignedBytesUnmovableShards.shardIds()
+                assignedBytesUnmovableShards.shardIds(),
+                unassignedBytesUnassignedShards.shards().size() > 0
```

Comment:
I wonder if we should compute the allocation explanation here when we already computed the allocation decision in `ReactiveStorageDeciderService.AllocationState#storagePreventsAllocation` and `ReactiveStorageDeciderService.AllocationState#storagePreventsRemainOrMove` maybe we could use that information? That should simplify things. wdyt @henningandersen?

Reply:
Yeah, we kind of do a double work here, but we also take advantage of the known JSON representation of `ShardAllocationDecision` that's used in the Cluster Allocation Explain API

Reply:
I am not sure we need to avoid the double work (is only for one shard), but I do think we should consider to not use the `AllocationService` to determine the result.

The `AllocationService` interacts with the gateway service and might provoke a fetch of shard data from all nodes. I'd prefer to decouple all that code here, since we do not rely on the gateway service for the determination of autoscaling.

Also, the `AllocationService` uses a potentially newer `ClusterState` and `ClusterInfo`. So the reason obtained may be different from the reason we decided to autoscale (or no reason at all). Additionally, `AllocationState.state` has been fixed wrt. shrink/split/clone shards to avoid infinite scale up. We should therefore use that cluster state to get the reasons to avoid confusion.

I think we can still use `AllocateUnassignedDecision` to produce the JSON. Or we can simply output the decisions including reasons one by one from the autoscaling code. We could reuse `NodeAllocationResult` to produce the output per node at least, perhaps using `AbstractAllocationDecision.nodeDecisionsToXContent`.

Also, a specific tier only concerns itself with the nodes in the tier. I think we could limit the output to only have decisions for those nodes rather than all nodes. This might not be entirely true for the move case, but I think it is for the unassigned case.

This would also avoid the dependency on `AllocationService`, simplifying this PR (and the code in general).

Reply:
Hey @henningandersen, thanks for the feedback! I've reworked the PR to get rid of `AllocationService` and bubble `NodeAllocationResult` from `cannotAllocateDueToStorage` and `storagePreventsRemainOrMove` up to the `scale` method, and output them in `ReactiveReason`.

Let me know what you think about it!